src/libs/actions/Session.js

```diff
@@ -96,6 +96,9 @@ function fetchAccountDetails(login) {
                     createAccount(login);
                 }
             }
+        })
+        .finally(() => {
+            Onyx.merge(ONYXKEYS.SESSION, {isLoading: false});
```

Comment:
This works, but I just want to make sure my original suggestion was known here. i.e. that we should consider linking the loading states with the data keys we are waiting to populate.

e.g. if we are calling `GetAccountStatus` then it would make sense to tie the loading state into `ONYXKEYS.ACCOUNT`. Just an idea. Not sure if it's super valuable to do now. But e.g. then we could skip this call to `finally()` and instead do:

```js
Onyx.merge(ONYXKEYS.ACCOUNT, { 
    loading: false,
    ...params,
});
```

Yeah this does mean we'd have to update all the places where this is used to be more specific instead of reusing `session` as a general loading state tracker, but I think it's a good pattern that if a component is calling an API it will be subscribing to the data it modifies as well.

Reply:
OK, I think that makes sense. Are you proposing that there would be both `account.isLoading` and `session.isLoading`? Is that all or would there also be `credentials.isLoading`?

Reply:
Yep all of them. In other words, we are waiting for some _data_ to load so the key where we will eventually set the data will hold the `loading` state.

Reply:
OK, so it would work like this:

1. Login form - `account.loading`
2. GitHub username form - this needs both `account` and `credentials`
3. Password form - also needs both `account` and `credentials`
4. Resend validation link form - also needs both `account` and `credentials`

Since three of them need access to two pieces of Onyx data, and it's the same data needed for every step, I'm not sure what Onyx key would be used for those cases.

Reply:
Good question and conversation. I guess I'm not sure how to answer. I want to say whatever data model you are updating with data returned from the API response would be what you toggle the `loading` state on. But I guess there could be multiple APIs being called and different data being set so maybe this isn't the best solution!

We are sort of replacing the traditional "make API request, handle response, do something" flow with something where the "done" state of the action is hooked into some random-ish piece of Onyx state. 

Going high level with this for a second, but what if every action had it's own meta data key in Onyx and we set these loading states _there_ instead of tying them to components (which seems bad for reusability) or sticking them on pieces of data (which maybe don't need loading states at all)

Reply:
> We are sort of replacing the traditional "make API request, handle response, do something" flow with something where the "done" state of the action is hooked into some random-ish piece of Onyx state.

I think this is a good thing to get clarity on. The way I see it, and also is the vision behind what @cead22 is attempting in https://github.com/Expensify/Expensify.cash/pull/915 is that basically the entire app is a reflection of what's in Onyx. There is a spectrum between nothing in Onyx (you are signed out and not looking at anything) to Onyx being fully loaded (you are looking at reports and adding actions). The UI just follows along on this journey that essentially is all about getting the right data into Onyx. 

So yes, it's definitely different than how we would traditionally approach this. I think that's why just using a single `session.loading` is fine because the entire sign on process is all about populating `session.authToken`.

I'm not really a big fan of letting actions store meta-data in their own Onyx keys because I think that can be easily abused. What happens when enough of that meta data uses up the storage limits and reports can't be stored anymore? I know that's extreme, but it could eventually happen.

I think that if we don't use `session.isLoading`, then I'd prefer to go back to using state in the individual forms (that code was more simple and used less lines to do the same thing). One thing that I've seen with testing is that the `isLoading` can persist between page refreshes if you refresh at the right time (or if you refresh when an API request is hanging) which actually makes having it be in Onyx a little bit worse in my opinion.

Reply:
>which actually makes having it be in Onyx a little bit worse in my opinion

Interesting take, but that does sound like a possible side effect of the data journey the UI is on. 

Is the persistence of the loading spinner because we are redirecting before the loading state has updated? If we redirect once our action resolves (rather than determining which view to display based on the state of our various props) then maybe we could avoid this? e.g.

- Make request A
- Handle response A
- Set loading state to false
- Set the correct view after the action is done

I'll admit, that doesn't fit the approach you are describing of a UI that is going on a data journey. But this is what the journey looks like now...
https://github.com/Expensify/Expensify.cash/blob/bf4926b93d884aa41d942a3e351485931cc5ebbc/src/pages/signin/SignInPage.js#L70-L106

Others might disagree, but this logic is somewhat difficult for me to parse.

Anyways, I won't block on these changes if you are OK with leaving the question of how to handle a loading state unresolved for now. I do think these types of questions will pop up in the future. I don't think `Onyx` provides clear guidance on how to tackle this yet. But it should eventually!

Maybe we can consider investigating / taking some inspiration from redux thunks in this case.
https://redux.js.org/tutorials/essentials/part-5-async-logic#loading-state-for-requests

Reply:
> Is the persistence of the loading spinner because we are redirecting before the loading state has updated? If we redirect once our action resolves

It doesn't have anything to do with redirects, no. There aren't any redirects in this flow at all. The situation is like this:

1. Submit the LoginForm
2. `isLoading` is now `true` and spinner on the button is shown
3. The API takes a long time to respond, maybe a couple of seconds
4. User gets impatient and refreshes the page
5. `isLoading` is still `true` so there is a spinner on the button even though nothing is loading

I think this kind of goes back to a convo I remember having with Rory about what is persisted to Onyx and what is not. I think one of the outcomes of that is that we should only put things on Onyx that we want to still be there after the page is reloaded. The open/closed state of the hamburger is a good example of something we want to be remembered after a refresh. I don't think loading spinners (especially ones associated with API requests that are canceled when the page is refreshed) are something that should be persisted.

This has me more convinced that this should be returned to the original implementation of using the state of the forms.

Reply:
After making that comment, I had another thought... 

Maybe we should have loading states automatically reset when the app loads. That could solve the problem I posted above.

As I looked into that, I came across the code in `Expensify.js` for `Onyx.init()` which actually does exactly that. Only it's clearing out `session.loading` and my code here is using `session.isLoading`.

Now there is a bigger issue with my code in this PR as those two properties are very confusing and not consistent that I need to think through.